Texture Cache: "Texture Groups" and "Texture Dependencies"

riperiperi opened the pull request:
## Intro

[![Download](https://img.shields.io/badge/Download-%232001-black?logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAARCAMAAADjcdz2AAAAe1BMVEVHcEyai5T/XFEAyOr/XlQBxeUBxeX9X1YDxOT/X1UBxeX7XlQAyev7X1QCxeX4YVj+X1UBxeRtmqhnnKv+X1UBxOQCxeUBxOT/X1UAzvELwd8Lwd//VUj+X1U2r8fYbmoAx+j/V0r/X1UCxeb/YlcCzO3/Z1wC1vkC0fPqHM+tAAAAInRSTlMACoFJz879AwP89FBlFeA7o5IZLN+tvZPdFRw0WrZsaX1CnD5B2QAAAI1JREFUGNNNz1cOwyAQBNABO2DAVek9ARvn/icMxSD282k0mkUzyWXXwB93B0wrSxAQ0mSgIx0AwxLwi7WzAssJXmsxHxIQQng9C6FhIlT7LxzoBNXnTTsPYoN7+xhITISOW/uiihfgKhQ2CDtwJV0Egd7IvCNA+1vWPoPbgep4OpeJJn/qwY5ACfrZ/QHYMQ4WY6nIhQAAAABJRU5ErkJggg==&style=flat&labelColor=252a2f&color=f2f5f8)](https://w.ryujinx.org/dl/2001)

Our texture cache tries to make sense of a low level collection of textures iteratively as draws and copies come in and attempt to use them. When _part_ of an existing texture is used, or an existing texture is used in another format, a _view_ of the existing texture is created, with the existing texture being the "storage". Existing textures can also _become_ views when a new texture creates them, with that new texture becoming storage. This enables a ton of fun things like writing to mip levels, texture arrays and cubemaps, as drawing to these views will draw to the original texture.

The key fact is, all of these textures _share their data_ through the storage texture. When the data for one view is set, all overlapping views will also see that change - they are linked at their core. Before now, views tracked their memory as if they were completely _separate_:
 - Before range-tracking, they would check for modified flags in a page table shared between all textures. This meant that a dirty flag could be cleared by another overlapping texture, and a different storage texture (or even views on the same page) wouldn't see that modified flag.
 - After range tracking, each view had its _own_ tracked handle, which covered its entire memory range. However, this meant that a dirty flag could be consumed on one view, but still present on another. Because synchronizing texture data for one view affects all overlapping ones, this is wrong too. Due to this aggressively setting texture data to its original state, there are a few workarounds to disable memory tracking for views, which cause problems for games that stream in or replace a lot of overlapping textures.
 
## Texture Groups

The real solution to these problems is to acknowledge the link between views of the same storage, and share all memory tracking between them. The structure used to represent a storage texture and its views is called a **Texture Group**, and acknowledges that a storage texture is made up of a number of sub-images, where each view into the storage will cover one or more of them.

So for each sub-image, we want to know if it has been "dirtied" by a guest memory write, and needs to be reloaded. We also want to know whether a sub-image has been modified, and flush out or preserve that data when possible. We don't want to consume dirty flags that we don't use, and we don't want to miss any from partial texture writes or that were not consumed by a previous view, so each view needs to work on a _range_ of these sub image handles, shared between all views of the storage texture.

However, creating tracking handles for every sub-image of every image would be very wasteful. Most textures will never have views of their mips/layers at all - they just contain static texture data that doesn't overlap with anything. So, textures have three separate levels of subdivision: 
 - One where the whole texture is covered by one handle,
 - One where one subdivision of the texture has a view (handle for each layer/face for 2d, and a handle for each mip level for 3d)
 - One where all layers and mip levels are tracked individually - there is a 1:1 mapping between view offsets and handles.
 
Because the layer/face view ordering changes when a texture is 3d, the presence of mip and layer views are tracked separately, and the difference in calculation happens when calculating tracking handles, and converting back to view indexes for the storage.

When additional views are created, the tracking will be subdivided further if needed, and the existing dirty/modified flags will be migrated onto the new handles.

#### Partial Synchronization
Since a view can cover multiple sub-images within the storage, and we have tracking (or potential modification) for each of these sub-images, it's possible that only a selection of them still have dirty flags. In most cases, we can still synchronize the full view, as it is faster to do than individual sub-images if the data on CPU is indeed the data we want. That is - in most cases.

Say that you have a texture array of 6 2d textures, and the first texture in the array is bound as a render target and rendered to. It consumes a dirty flag and synchronizes that first 2d array item from CPU memory - memory that the game cleared for us before use - then renders to it and flags it as GPU modified. So far so good, we have data in the first and no data in the other 5. But what if we suddenly bind the full texture array? When we check the dirty flags for the 6 overlapping faces of the storage, the first is consumed, but the other 5 are not. Synchronizing the full texture would clear out the data we just rendered into the first 2d item. We know that one of the 6 textures has been modified and not flushed out, so we'll need to _partially sync texture data_ from the CPU to honor the 5 other dirty flags.

To do this, we keep track of which of the overlapping handles had a dirty flag. If _at least one, but not all of them_ had a dirty flag, AND at least one of the non-dirty textures is gpu modified, then we need to selectively sync _only_ the textures with the dirty flags to avoid losing the data for the texture we just wrote.

A few changes were made to achieve this. There is an additional method to set a single 2D slice of a texture on the backend, and the `ConvertToHostCompatibleFormat` function has been updated to take a `level` and optionally convert layout for just a single level/layer, rather than all of them. And finally, this is the reason that `Texture.SynchronizeMemory` doesn't do any synchronizing itself, it calls the `TextureGroup` to decide what to do _first_. This will also help in future for other things, and is made just as fast as before by the per texture `_dirty` flag.

### Additional Changes:
A few changes were made so that the assumptions made by the Texture Group were correctly held:
- RegionHandles now let you register an event that is fired when "Dirty" becomes true. This cannot be unregistered, it's assumed that this callback will be relevant until the handle is disposed.
- RegionHandles can now be reprotected as dirty. This is used for inheriting dirty flags from another handle.
- When a texture with views is converted into a view itself, its views will be migrated to the new storage texture first.
  - This may be responsible for improving gpu memory usage in HW:AOC
  - This may fix some other unusual errors, though the process of "stealing" views is still a little flawed without a copy dependency.

### Fixes:
- Texture swaps in Xenoblade Chronicles: Definitive Edition, and Xenoblade 2
  - May fix texture swaps in other games
- Texture memory leak in Hyrule Warriors: Age Of Calamity (possibly additional change 1)
- Small improvements in A Hat in Time and Mario Kart 8, fixed further by what you see further down in the comments.
- Greatly reduced number of texture tracking handles (and frequency of their creation) when views are created. This will spend less time creating the handles, which locks the tracking structure.

## TODO: Future Texture Changes
None of these issues have anything to do with this PR, I just wanted to mention them to manage expectations (these things are not yet fixed), and to plan out a roadmap for future changes of this kind.

- Texture eviction strategy that keeps textures alive (after falling off the pool) in case they were simply moved to another location in the pool, or temporarily removed.
  - Stutters and performance in Hyrule Warriors: Age Of Calamity
- Texture eviction strategy that evicts textures when an overlapping texture is created that is not compatible in any way (its data will be garbage)
  - VRAM usage in Xenoblade Chronicles: Definitive Edition (it's also a bit slower cos of buffer flush, but that's not related at all)
- Keep textures _in the cache_ when a view is still alive.
 - Textures are still kept _alive_ when any view is still alive, but they are not findable via the texture cache. This can lead to situations where we end up recreating a storage texture that already exists, or creating a recursive view train between formats, and doing a bunch of unnecessary work over time, and adding a bunch of confusion when finding overlaps which could impact changes.

#### Texture Dependencies (actually done, see below)
Texture views only perfectly represent the guest state if a texture can only have one storage. In some cases, a texture may end up having separate potential "storages", due to the following:
 - View texture compatible with two or more disjoint storage layouts (mario kart - offset overlapping texture arrays + cubemaps)
 - View cannot be created due to host incompatibility, such as compressed views of uint textures (mario kart, a hat in time, splatoon 2)
 
This change is made a lot more possible by the changes in this PR. Dependencies would be _between groups_, as they are at their core an attempt to keep two overlapping storage textures in sync.

## Testing requests
You can download a build of this PR here: https://w.ryujinx.org/dl/2001

Please test everything. I wish I was kidding. Particular attention to:

- Resolution Scaling. This will definitely affect texture blacklist rules, so make sure all your fave games still work.
- Games which load and stream a lot of textures. (UE4)

[![Download](https://img.shields.io/badge/Download-%232001-black?logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAARCAMAAADjcdz2AAAAe1BMVEVHcEyai5T/XFEAyOr/XlQBxeUBxeX9X1YDxOT/X1UBxeX7XlQAyev7X1QCxeX4YVj+X1UBxeRtmqhnnKv+X1UBxOQCxeUBxOT/X1UAzvELwd8Lwd//VUj+X1U2r8fYbmoAx+j/V0r/X1UCxeb/YlcCzO3/Z1wC1vkC0fPqHM+tAAAAInRSTlMACoFJz879AwP89FBlFeA7o5IZLN+tvZPdFRw0WrZsaX1CnD5B2QAAAI1JREFUGNNNz1cOwyAQBNABO2DAVek9ARvn/icMxSD282k0mkUzyWXXwB93B0wrSxAQ0mSgIx0AwxLwi7WzAssJXmsxHxIQQng9C6FhIlT7LxzoBNXnTTsPYoN7+xhITISOW/uiihfgKhQ2CDtwJV0Egd7IvCNA+1vWPoPbgep4OpeJJn/qwY5ACfrZ/QHYMQ4WY6nIhQAAAABJRU5ErkJggg==&style=flat&labelColor=252a2f&color=f2f5f8)](https://w.ryujinx.org/dl/2001)

gdkchan commented on Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:
```diff
@@ -0,0 +1,669 @@
+                    if (handle.Dirty)
+                    {
+                        handle.Reprotect();
+                        dirty |= true;
```
Same comment applies here too.

gdkchan commented on Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:
```diff
@@ -0,0 +1,669 @@
+        /// Synchronize part of the storage texture, represented by a given range of handles.
+        /// Only handles marked by the _loadNeeded array will be synchronized.
+        /// </summary>
+        /// <param name="baseRegion"></param>
```
Missing description for parameters.

gdkchan commented on Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:
```diff
@@ -0,0 +1,669 @@
+        /// Evaluate the range of tracking handles which a view texture overlaps with.
+        /// </summary>
+        /// <param name="texture">The texture to get handles for</param>
+        /// <returns></returns>
```
Missing.

gdkchan commented on Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:
```diff
@@ -0,0 +1,669 @@
+                    // Only one of these will be greater than 1, as partial sync is only called when there are sub-image views.
+                    for (int layer = 0; layer < info.layers; layer++)
+                    {
+                        for (int level = 0; level < info.levels; level++)
```
Looking at this made me realise that theres a bug on the way `allOffsets` length is calculated right now, it doesn't take into account the fact that the layers count halve for 3D textures at each mip level... So this will probably read the wrong offset (zeros) while iterating it (for 3D textures with levels > 1).

riperiperi replied:
Will be resolved in a future commit.

gdkchan commented on Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:
```diff
@@ -0,0 +1,669 @@
+        /// <returns>Start index and count of offsets for the given level</returns>
+        private (int index, int count) Get3DLevelRange(int level)
+        {
+            int index = 0;
```
Why not just:
```cs
int index = _layers * level;
int count = Math.Max(_layers >> level, 1);
```

riperiperi replied:
Each level should have less layers than the last. Imagine a 3d texture with a depth of 4, and 4 mip levels:

```
[
1, 2, 3, 4, 5, 6, 7, 8,
9, 10, 11, 12,
13, 14,
15
]
```
Calculating the index like that will result in an incorrect offset index for the second mip. The count needs to be reduced each time so it's like (8 + 4 + 2) etc.

gdkchan replied:
Ah ok. Its worth noting it isn't actually stored like that on the `AllOffsets` array right now, as I mentioned before.

gdkchan commented on Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:
```diff
@@ -0,0 +1,669 @@
+
+                    int layerLevels = _levels;
+
+                    while (handleIndex >= layerLevels)
```
You can get rid of this loop by taking the difference, and then using multiplication, shift, etc.

gdkchan commented on Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:
```diff
@@ -0,0 +1,669 @@
+        /// </summary>
+        /// <param name="handle">The handle this flush action is for</param>
+        /// <param name="address">The address of the flushing memory access</param>
+        /// <param name="address">The size of the flushing memory access</param>
```
```suggestion
        /// <param name="size">The size of the flushing memory access</param>
```

gdkchan commented on Ryujinx.Graphics.Gpu/Image/TextureInfo.cs:
```diff
@@ -232,6 +232,27 @@ public static int GetLayers(Target target, int depthOrLayers)
         }
 
+        /// <summary>
+        /// Gets the number of 2d slices of the texture.
```
```suggestion
        /// Gets the number of 2D slices of the texture.
```

gdkchan commented on Ryujinx.Memory/Tracking/RegionHandle.cs:
```diff
@@ -120,11 +138,22 @@ public bool OverlapsWith(ulong address, ulong size)
             return Address < address + size && address < EndAddress;
         }
 
+        bool disposed;
```
I think this should be called `_disposed` and moved next to the other fields.

riperiperi replied:
This is actually a leftover from testing, where a double dispose was causing issues. I didn't plan to make it into an actual change.

riperiperi replied:
Though, throwing an `ObjectDisposedException` would make it obvious if this error were encountered again.

gdkchan commented on Ryujinx.Memory/Tracking/RegionHandle.cs:
```diff
@@ -120,11 +138,22 @@ public bool OverlapsWith(ulong address, ulong size)
         {
+            if (disposed)
+            {
+                throw new Exception("Double Dispose!");
```
```suggestion
                throw new ObjectDisposedException();
```

gdkchan commented on Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:
```diff
@@ -0,0 +1,669 @@
+        /// </summary>
+        /// <param name="handleIndex">The index of the handle</param>
+        /// <returns>The layers and levels that the handle covers, and its index in the offsets array</returns>
+        private (int baseLayer, int baseLevel, int levels, int layers, int index) GetHandleInformation(int handleIndex)
```
`(int BaseLayer, int BaseLevel, int Levels, int Layers, int Index)`

gdkchan commented on Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:
```diff
@@ -0,0 +1,669 @@
+        /// </summary>
+        /// <param name="level">The level to return</param>
+        /// <returns>Start index and count of offsets for the given level</returns>
+        private (int index, int count) Get3DLevelRange(int level)
```
`(int Index, int Count)`

gdkchan commented on Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:
```diff
@@ -0,0 +1,669 @@
+        /// </summary>
+        /// <param name="texture">The texture to get handles for</param>
+        /// <returns></returns>
+        private (int baseHandle, int regionCount) EvaluateRelevantHandles(Texture texture)
```
`(int BaseHandle, int RegionCount)`

gdkchan reviewed (commented):
Great work, Xenoblade has never looked so good. I will test some games later.

Illynir commented:
No problem on FETH, Buried Stars and MHR Demo.

Xenoblade 1 and 2 are almost perfect now.

Leischii commented:
Already mentioned on Discord but to make sure it's tested once fixed: AC-NH crashes with this after the daily announcement.

scese250 commented:
No regression found for Girls und Panzer DTM DX. 

No changes for DOA Xtreme 3 Scarlet unfortunately.

ReaperOfSouls1909 commented:
No regressions in ue4 titles and nothing improved 

Sent from my iPhone

> On Feb 8, 2021, at 6:45 PM, scese250 <notifications@github.com> wrote:
> 
> ﻿
> No regression found for Girls und Panzer DTM DX.
> 
> No changes for DOA Xtreme 3 Scarlet unfortunately.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

CryptForge commented:
Age of Calamity appears to gain a small increase in performance and better stability with this PR. Especially when resolution is scaled up.

LukeWarnut commented:
I found a regression in 20XX with this PR.

Before PR:
![image](https://user-images.githubusercontent.com/65521430/108570347-a5ccd180-72db-11eb-916c-e8988272ccf9.png)
[20XX.log](https://github.com/Ryujinx/Ryujinx/files/6013459/20XX.log)

After PR:
![image](https://user-images.githubusercontent.com/65521430/108570368-b4b38400-72db-11eb-911a-c35e05e87bca.png)
[20XX-2001.log](https://github.com/Ryujinx/Ryujinx/files/6013442/20XX-2001.log)

gdkchan commented:
Can you include steps to reproduce the issue? And also a save if necessary/possible.

LukeWarnut commented:
> Can you include steps to reproduce the issue? And also a save if necessary/possible.

Those screenshots were taken in the hub world after beating the tutorial. (Full screenshot below)
Here's my save file that has the tutorial completed:
[0.zip](https://github.com/Ryujinx/Ryujinx/files/6013541/0.zip)
![image](https://user-images.githubusercontent.com/65521430/108570987-1b856d00-72dd-11eb-9773-fbc0d117f2ec.png)

riperiperi commented:
Don't review right now, there are more changes to come.

riperiperi commented:
# Copy Dependencies

Copy Dependencies are an extension of the "CopyOnly" compatibility that was previously used to copy 3D texture slices to the full 3D texture on creation, and the emulated views (copy driven) we used for compressed<->uncompressed views, but in a more general sense.

Firstly, the CopyOnly compatibility has been extended to include view-parent pairs with mismatching compressed/uncompressed formats. It is also set when a texture is to be taken as view of a new texture, but it is already a view of another, as we cannot have a texture with two storage textures.

Before, CopyOnly compatibility would only be honored at the time of texture creation, and _only_ when a storage texture was created, consuming the copyable views. Updates to these "views" would not cause future copies to the parent. Because of this, there were a number of workarounds that attempted to avoid losing this initial data by synchronizing from memory, as future draws to update it (presumably after a cpu memory clear) would be lost.

To solve this problem, we use the flexibility of Texture Group Handles to implement "copy-dependencies" - a sort of contract that indicates that a write to one sub-image of the storage texture _must_ result in a copy to a sub-image of another texture before the next time it is used. In the context of Texture Groups, the dependency is between two TextureGroupHandles, and copies are deferred to dependent textures when SignalModified is called. This sets the `_dirty` flag on overlapping textures, which will indicate that it should check its dependencies to see if 

This fits nicely in with the partial texture updates added by Texture Groups. Deferred copies are only performed if their source handle has not been dirtied after its last use, and will override dirty flags on the side of the copy target (as due to the copy source dirty flags being from the same memory region, the copy will be newer). It's even possible for multiple copies to be performed _alongside_ synchronizing several sub-images of a texture from memory. The BBC will have to do better than strikingly similar nature documentaries narrated by a guy with a nice voice to top how beautiful this all is.

Note that creating a copy dependency forces the texture group to use the most granular handle subdivision, as the copies happen for a single 2D sub-image.

## Additional Fixes

- Tracking can no longer be disabled on a texture.
- View compatibility now also checks:
  - If the levels/layers of the view are in bounds of the parent. If they are not, the texture is incompatible.
  - If the view has more than 1 slice, its LayerSize must match the parent's (otherwise only the base layer is a view).
- As a result of copy dependencies doing the work instead, the emulated view parent has been removed.
- Fixed many issues with 3D texture views.
  - View offsets are now properly calculated for 3D textures with mipmaps.
  - View offsets are now properly transformed into layer/level for 3D textures. (also, the all offsets binary search now works correctly)
  - Fixed some issues with 3D textures and Texture Groups.
- Render targets now notify the texture when it is _bound_, and _unbound_. Copies performed while a texture is bound are repeated every time, rather than being cleared, as the texture is still being written to.
- Other minor rearrangements and fixes.

# Comparisons

Screenshots were taken with a build before some final changes.

## Mario Kart 8 Deluxe
- Fixed majority of missing faces on ingame cubemaps (4800% more cubemap minimum)
- Fixed issue where characters/karts would become partially or fully black after cycling through character select for a while.
- Fixed colour levels being incorrect for most courses. (rendered 3D texture LUT)
(there is still an issue where certain interior lighting can infrequently cause a rainbow effect that is being looked into, and one face of a cubemap's mip layers may be lost past the first race.)

### Before
Character Select:
![](https://cdn.discordapp.com/attachments/157873776040607744/812751013135122462/unknown.png)
Electrodrome:
![](https://cdn.discordapp.com/attachments/157873776040607744/812751547284979722/unknown.png)
![](https://cdn.discordapp.com/attachments/157873776040607744/812751578620100608/unknown.png)
Royal Raceway:
![](https://cdn.discordapp.com/attachments/157873776040607744/812751187673088020/unknown.png)
![](https://cdn.discordapp.com/attachments/157873776040607744/812751370939662366/unknown.png)
### After
Character Select (source: just trust me bro):
![](https://cdn.discordapp.com/attachments/157873776040607744/812749167856189510/unknown.png)
Electrodrome:
![](https://cdn.discordapp.com/attachments/157873776040607744/812749918774099978/unknown.png)
![](https://cdn.discordapp.com/attachments/157873776040607744/812749927746502735/unknown.png)
Royal Raceway:
![](https://cdn.discordapp.com/attachments/157873776040607744/812749290740908052/unknown.png)
![](https://cdn.discordapp.com/attachments/157873776040607744/812749703485063258/unknown.png)

## A Hat in Time
- Majority of texture issues fixed. Some texture related crashes fixed.
- Game still devolves into madness after a while of playing... but it's a lot better.

### Before
Title Screen:
![](https://cdn.discordapp.com/attachments/157873776040607744/812727325476257832/unknown.png)
Mafia Town:
![](https://cdn.discordapp.com/attachments/728011297857339422/812729810324684880/unknown.png)
...just kidding (3 crashes later):
![](https://cdn.discordapp.com/attachments/728011297857339422/812731341317472266/unknown.png)

### After
Title Screen:
![](https://cdn.discordapp.com/attachments/728011297857339422/811310805382987806/unknown.png)
Mafia Town:
![](https://cdn.discordapp.com/attachments/728011297857339422/811312460882903040/unknown.png)

## Super Mario Odyssey
- Fixed colour levels being incorrect for most kingdoms. (rendered 3D texture LUT)
- Fixed various issues caused by a rendered mipmapped 3d perlin noise texture. (that's a lot of words)
  - Fog in all kingdoms is now wispy, rather than filled with harsh lines and unrealistic changes in intensity.
  - Fixed all effects relating to the Seaside Kingdom water. (beach waves, caustics, surface colour)
  - Fixed "dirty" decals that are applied to mario in certain situations. (smoke, fog)

### Before
Fog (metro):
![](https://cdn.discordapp.com/attachments/157873776040607744/812378750435196938/unknown.png)
Fog (wooded):
https://gyazo.com/0f4c0c1e8084e6b491e7a05f11c37bef
Seaside Waves:
https://gyazo.com/f9133048bf9173085ea52ce25850faa1

### After
Fog (metro):
![](https://cdn.discordapp.com/attachments/157873776040607744/812377545281634364/unknown.png)
Fog (wooded):
https://gyazo.com/0ef88252830c80f8c94d59051d3a8570
Seaside Waves:
https://gyazo.com/d757c516ba23bc3355e9c4edf33d5639
Seaside Caustics:
https://gyazo.com/6b3b9d1b83dbb68150c21c18eb305f4b

## Splatoon 2
- Fixed an issue where the plaza would be really dark on the first load, or randomly on future loads/levels.
- Fixed issues where text on the large splatter buttons would not appear, or would be incorrect. (this also affects large text in multiplayer)
- Fixed colour levels being incorrect occasionally. (rendered 3D texture LUT)

### Before
Plaza (1st load):
![](https://cdn.discordapp.com/attachments/157873776040607744/812753766218203198/unknown.png)
Menu buttons:
![](https://cdn.discordapp.com/attachments/157873776040607744/812754152068350022/unknown.png)
### After
Plaza (1st load):
![](https://cdn.discordapp.com/attachments/728011297857339422/811302658971402260/unknown.png)
Menu buttons:
![](https://cdn.discordapp.com/attachments/157873776040607744/812754930099290142/unknown.png)

## Xenoblade Chronicles: Definitive Edition (maybe 2?):
- Fixed colour levels being subtly incorrect. (rendered 3D texture LUT, every frame)

### Before
![](https://cdn.discordapp.com/attachments/728011297857339422/812037714564087838/unknown.png)
### After
![](https://cdn.discordapp.com/attachments/728011297857339422/812038151014055946/unknown.png)

There might be a better example somewhere else.

## Animal Crossing: New Horizons
- Fixed DOF only using the first blur level, rather than all of them. (a copy each frame is required to do this)
  - Only really noticeable for very far away, or cutscenes that enforce a heavy DOF. Pay attention to the museum at the back.
### Before
Pay attention to the museum.
https://cdn.discordapp.com/attachments/728011297857339422/811957312377585694/unknown.png
### After
Pay attention to the museum.
https://cdn.discordapp.com/attachments/728011297857339422/811957328995024936/unknown.png

## Luigi's Mansion 3:
- May have fixed some lighting or room transition related issues.

## Pikmin 3 Deluxe:
- Fixed intro video showing a black screen.
- May fix subtle issues with the KopPad.

### After (black screen before)
![](https://cdn.discordapp.com/attachments/157873776040607744/812104072735096832/unknown.png)

May fix similar issues with other Nintendo first party games. Changes should be tested with games with texture issues.

# Note

This isn't a perfect solution, but it has the potential to be one. I'll be looking into the weird remaining issues, but what I have right now is very reviewable and should work better in most cases.

legend800 commented:
Fixes missing menu text in Fast RMX.

riperiperi commented:
> 
> 
> > Can you include steps to reproduce the issue? And also a save if necessary/possible.
> 
> Those screenshots were taken in the hub world after beating the tutorial. (Full screenshot below)
> Here's my save file that has the tutorial completed:
> [0.zip](https://github.com/Ryujinx/Ryujinx/files/6013541/0.zip)
> ![image](https://user-images.githubusercontent.com/65521430/108570987-1b856d00-72dd-11eb-9773-fbc0d117f2ec.png)

Can you retest this with the latest commit? I'm not too sure what would cause it.

Niwu34 commented:
Master:
![20XX - Master](https://user-images.githubusercontent.com/67392333/108807549-0cd0db80-75a5-11eb-86e3-f89e13b586b3.PNG)

Pr 2001 []Ryujinx 1.0.6668:
![20XX - Texture](https://user-images.githubusercontent.com/67392333/108807600-1eb27e80-75a5-11eb-8878-165379cd4047.PNG)

riperiperi commented:
Latest commit seems to fix remaining issues with Mario Kart 8: "and one face of a cubemap's mip layers may be lost past the first race."

Shahil-Ayato commented:
Fixes #1620

Shahil-Ayato commented:
Could this potentially fix #1915?

riperiperi commented:
> 
> 
> Could this potentially fix #1915?

Yes.

riperiperi commented on Ryujinx.Graphics.Gpu/Image/TextureCompatibility.cs:
```diff
@@ -382,10 +403,22 @@ public static bool ViewLayoutCompatible(TextureInfo lhs, TextureInfo rhs, int le
         /// <param name="rhs">Texture information of the texture view</param>
-        /// <returns>True if the formats are compatible, false otherwise</returns>
-        public static bool ViewFormatCompatible(TextureInfo lhs, TextureInfo rhs)
+        /// <returns>How view compatible the formats are</returns>
```
Yes, it just reads rather unusually since it starts with "how". The returned value is a measure of the view compatibility of the formats, so the function returns how "view compatible" the formats are. I can think of another way to word it.

riperiperi commented on Ryujinx.Graphics.Gpu/Image/TextureManager.cs:
```diff
@@ -841,24 +943,27 @@ public Texture FindOrCreateTexture(TextureSearchFlags flags, TextureInfo info, i
                 {
                     Texture overlap = _textureOverlaps[index];
+
+                    if (overlap.Info.Target == Target.Cubemap && texture.Info.Target == Target.CubemapArray) { }
```
It just keeps happening 🥴

riperiperi commented on Ryujinx.Graphics.Gpu/Image/Texture.cs:
```diff
@@ -544,18 +590,46 @@ public void SynchronizeMemory()
             if (_hasData)
             {
-                if (_memoryTracking?.Dirty != true)
+                if (Group != null)
```
`Group` actually isn't set null anymore. I'll change this and ConsumeModified to reflect that.

riperiperi commented on Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:
```diff
@@ -0,0 +1,985 @@
+        /// </summary>
+        /// <param name="level">The level of the view</param>
+        /// <returns>The view's size in bytes</returns>
+        private int CalculateViewDataSize(int level)
```
I've addressed this by making the size calculator output an array of `Slice Sizes` for each mip level, which can be indexed by level rather than using this function. This is given to the texture group as part of initializating the storage.

riperiperi commented:
Should be all comments addressed. There shouldn't be any regressions, but it's worth testing anyways.

gdkchan commented on Ryujinx.Graphics.Gpu/Image/TextureCompatibility.cs:
```diff
@@ -235,6 +235,27 @@ public static TextureViewCompatibility ViewSizeMatches(TextureInfo lhs, TextureI
+        /// <param name="child">Texture information for the child</param>
+        /// <param name="layer">Base layer of the child texture</param>
+        /// <param name="level">Base level of the child texture</param>
+        /// <returns>Full compatiblity if the child's layer and level count fit within the parent, incompatibile otherwise</returns>
```
Typo: `incompatibile`.

gdkchan commented on Ryujinx.Graphics.Gpu/Image/TextureDependency.cs:
```diff
@@ -0,0 +1,36 @@
+        /// The handle that owns this dependency.
+        /// </summary>
+        public TextureGroupHandle Handle;
+        /// <summary>
```
Missing empty line separation before this line.

gdkchan commented on Ryujinx.Graphics.Gpu/Image/TextureGroupHandle.cs:
```diff
@@ -0,0 +1,327 @@
+        /// <summary>
+        /// Signal that a copy dependent texture has been modified, and must have its data copied to this one.
+        /// </summary>
+        /// <param name="copyFrom">The texture handle that must defer a copy to this one.</param>
```
```suggestion
        /// <param name="copyFrom">The texture handle that must defer a copy to this one</param>
```

gdkchan reviewed (approved):
lgtm other than the few comments.
Awesome work fixing all those bugs 👍

michaelpegaso commented:
With the latest updates it has stopped having effect, for example in splatton 2 the problem of shadows in the initial square of the game is once again.

Leischii commented:
> 
> 
> With the latest updates it has stopped having effect, for example in splatton 2 the problem of shadows in the initial square of the game is once again.
> 
> I have tried the "Output Windows x64 (Release)" version and everything seems to work perfect.

What do you mean by "latest updates"? Did you use the built in updater? If that's the case, then the changes made in this Pr have been overwritten with latest master, which doesn't have these changes.

Thog reviewed (approved):
lgtm
